modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbMvPartitionStorage.java

```diff
@@ -0,0 +1,496 @@
+import org.apache.ignite.internal.storage.StorageException;
+import org.apache.ignite.internal.storage.TxIdMismatchException;
+import org.apache.ignite.internal.tx.Timestamp;
+import org.apache.ignite.internal.util.Cursor;
+import org.apache.ignite.internal.util.GridUnsafe;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.lang.IgniteInternalException;
+import org.jetbrains.annotations.Nullable;
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.ReadOptions;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+import org.rocksdb.Slice;
+import org.rocksdb.WriteOptions;
+
+/**
+ * Mult-versioned partition storage implementation based on RocksDB. Stored data has the following format:
+ * <pre><code>
+ * | partId (2 bytes, BE) | rowId ({@link #igniteRowIdSize} bytes) |</code></pre>
+ * or
+ * <pre><code>
+ * | partId (2 bytes, BE) | rowId ({@link #igniteRowIdSize} bytes) | timestamp (16 bytes, DESC) |</code></pre>
+ * depending on transaction status. Pending transactions data doesn't have a timestamp assigned.
+ *
+ * <p/>BE means Big Endian, meaning that lexicographical bytes order matches a natural order of partitions.
+ *
+ * <p/>DESC means that timestamps are sorted from newest to oldest (N2O). Please refer to {@link #putTimestamp(ByteBuffer, Timestamp)} to
+ * see how it's achieved. Missing timestamp could be interpreted as a moment infinitely far away in the future.
+ */
+public class RocksDbMvPartitionStorage implements MvPartitionStorage {
+    /** Position of row id inside of the key. */
+    private static final int ROW_ID_OFFSET = Short.BYTES;
+
+    /** Timestamp size in bytes. */
+    private static final int TIMESTAMP_SIZE = 2 * Long.BYTES;
+
+    /** Maximum possible size of the key. */
+    private static final int MAX_KEY_SIZE = /* partId */ ROW_ID_OFFSET + /* rowId */ MAX_ROW_ID_SIZE + /* timestamp */ TIMESTAMP_SIZE;
+
+    /** Threadlocal direct buffer instance to read keys from RocksDB. */
+    private static final ThreadLocal<ByteBuffer> MV_KEY_BUFFER = withInitial(() -> allocateDirect(MAX_KEY_SIZE).order(BIG_ENDIAN));
+
+    /** Threadlocal ob-heap byte buffer instance to use for key manipulations. */
+    private final ThreadLocal<ByteBuffer> heapKeyBuffer;
+
+    /** Size of the {@link IgniteRowId} identifiers for this particular partition storage. */
+    private final int igniteRowIdSize;
+
+    /**
+     * Partition ID (should be treated as an unsigned short).
+     *
+     * <p/>Partition IDs are always stored in the big endian order, since they need to be compared lexicographically.
+     */
+    private final int partId;
+
+    /** RocksDb instance. */
+    private final RocksDB db;
+
+    /** Partitions column family. */
+    private final ColumnFamilyHandle cf;
+
+    /** Write options. */
+    private final WriteOptions writeOpts = new WriteOptions().setDisableWAL(true);
```

Comment:
Why is WAL being disabled here? I think the reason is worth mentioning in a comment.

Reply:
I'll remove this call for now. Storage state recovery procedure is not yet specified properly, so it's too early to disable WAL